### ISSUE TITLE
Stdlib: add Vector module (resizable arrays)

### ISSUE THREAD

**Issue by meadofpoetry** [ocaml/ocaml#9553]
It would be nice to have resizable arrays in Stdlib. [Containers' `CCVector` could be a good reference.](https://github.com/c-cube/ocaml-containers/blob/master/src/core/CCVector.mli)

**Comment by dra27** [comment]
A code-search on GitHub for `CCVector` and `BatVect` as two quick examples doesn't suggest it's _that_ widely used? It's not something with an immediately clear underlying representation either (`CCVector` is very similar to `Buffer`, creating a new underlying array when it's not big enough, `BatVect` is based on ropes).

I'm not at this stage convinced...

**Comment by meadofpoetry** [comment]
> A code-search on GitHub for CCVector and BatVect as two quick examples doesn't suggest it's that widely used?

Well yes, I suppose many people use lists for such cases. Or this may be due to not that many projects use Containers or Batteries.

> It's not something with an immediately clear underlying representation either

I was thinking about cache-friendly data structure with relatively cheap (amortized) element addition, so something akin to Python's list or CCVector, rather than ropes.

But you may be right, I've found only a couple of places where CCVecror is used, so maybe it doesn't worth it.

**Comment by dra27** [comment]
I looked at `CCVector` relative to say `CCList` and `BatVect` relative to `BatList` (both libraries are used extensively). I can certainly see that the containers-style version is useful (and very similar to `Buffer`), but I don't think it meets the "bar" for being supported in the standard library.

**Comment by bluddy** [comment]
I would argue that Vector as in CCVector is an extremely versatile data structure, which is why languages like python include it by default as their simplest data type. The reason it isn't more used in the OCaml world IMO is because people don't have access to it. CCVector provides:

* O(1) amortized push (append to the end of the data structure) complexity. This is not possible with lists or arrays.
* Near-zero allocation insert-anywhere (amortized).
* O(1) `length` operation (available on array, but not on list).

Basically, it's an excellent data structure once you want performance higher than what persistent data structures can provide, and providing easy access to it should have the benefit of speeding up OCaml's most time-sensitive libraries.

**Comment by XVilka** [comment]
A good examples are:
- Rust std::vec::Vec https://doc.rust-lang.org/std/vec/struct.Vec.html
- Haskell Vector (this one is an external library though) https://wiki.haskell.org/Numeric_Haskell:_A_Vector_Tutorial#Generating_Vectors

**Comment by tperami** [comment]
Among the libraries than need to be checked for usage, there is probably [res](https://opam.ocaml.org/packages/res/). `Res.Array` is a dynamically resizable array like `CCVector` or `BatVect`. It's a dependency of Jane Street `core` which maybe expose it under a different name.

**Comment by mariojppereira** [comment]
Let me just point out that in the [VOCaL](https://github.com/ocaml-gospel/vocal) library we also have an implementation of resizable arrays: https://github.com/ocaml-gospel/vocal/blob/main/src/Vector.mli

Moreover, we use the [Why3](https://gitlab.inria.fr/why3/why3) framework to formally verify the correctness of such implementation.

**Comment by gasche** [comment]
I tend to agree that growable/resizable arrays are a common enough data structure that it would be nice to have in the standard library. A few weeks ago I commented on the growth/resizing logic of a reimplementation [inside the ocaml-multicore library runtime](https://github.com/ocaml-multicore/ocaml-multicore/pull/582#discussion_r650758187), and it would have been nice to not have to do this at all.

The difficulty with such PRs is to converge on an API that everyone likes, and is a good fit for the stdlib.

**Comment by c-cube** [comment]
I just re-found this issue, but another critical thing is that by making `Vec.t` standard, the library can rely on OCaml's internal invariants to be more optimized. (I feel like it's discussed in another similar issue?). A big flaw of doing Vec in userland is that you can't have a partially uninitialized array, which makes the API very unergonomic compared to vectors in all the other languages. 

As for usefulness, @dra27, judging of the utility of non standard containers by usage is not great, I think. It's like saying that HAMTs are not so great because, in OCaml, they're rare; but convenience and interoperability are important factors, and clojure/scala users would be unhappy if they lost their HAMTs (or evolutions thereof). Similarly, many many imperative languages have vector and hashmap as their two core datastructures, and it's not by coincidence. Since OCaml is imperative, I think we miss half of the equation :)

**Comment by Octachron** [comment]
There is still work on progress at #11882 .

**Comment by yawaramin** [comment]
Looks like this can be closed now since Dynarray is merged?

**Comment by gasche** [comment]
Right, thanks for the triaging help. #11882 was merged, the Dynarray module will be available in OCaml 5.2.
